### PR TITLE
[java] New Rule: Use Explicit Types

### LINKED ISSUE #2847: [java] New Rule: Use Explicit Types
**Proposed Rule Name:** UseExplicitTypes

**Proposed Category:** One of [Best Practices | Error Prone]

**Description:**

Java 10 introduced the `var` keyword.  This reduces the amount of typing but decreases the reading comprehension of the code.  Please create a rule that flags the use of the var keyword.

**Code Sample:**

The line with `var index` should be flagged as a problem.

```
public static void myMethod()
{
   var index = 0;
}
```

Here is the "correct" code.
```
public static void myMethod()
{
   int index = 0;
}
```

**Possible Properties:**

* localVariables: One of [allow | disallow].  Defaults to `disallow`.
* forLoopIndexes: One of [allow | disallow].  Defaults to `disallow`.
* tryWithResources: One of [allow | disallow].  Defaults to `disallow`.
* primitiveTypes: One of [allow | disallow].  Defaults to `disallow`.  If `true`, then `var` can replace a primitive type as long as one of the above properties allows the location.

Are there any other places where `var` can be used?  If so, add those as properties.

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Most basic implementation of #2847.

Please review description and properties...

I've not added it to the quickstart.xml ruleset, as I think, this rule is a bit controversial and depends on the project.

I've chosen ~"Best Practices"~ "Code Style" as category and not error prone, as using `var` doesn't break the code, it's "just harder to read".

## Related issues

- Fixes #2847

## Ready?

- [x] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 5 violations,<br />
introduces 35 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/4a64b7bc1a764788857026960f20bb56/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 1]
```diff
@@ -1688,6 +1688,34 @@ public class Foo {
...
+    <rule name="UseExplicitTypes"
+          language="java"
+          minimumLanguageVersion="10"
+          since="7.0.0"
+          message="Use Explicit Types"
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#useexplicittypes">
+        <description>
+Java 10 introduced the `var` keyword. This reduces the amount of typing but decreases the reading comprehension of the
+code.
```
[the remark this answers is not in the document]
Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/bestpractices.xml:1700]: I would also moderate the blanket claim that it decreases reading comprehension, as we don't have substantial evidence that it does.

**Comment by oowekyala** [comment]
Thanks!
I think this rule is a better fit for the codestyle category though. 
[As the java devs put it](https://openjdk.org/projects/amber/guides/lvti-style-guide), explicit types are a trade-off. Sometimes type Inference makes code more readable, and sometimes it's the other way around. Disallowing `var` is a pretty extreme stance on the issue and I don't think it's a widely accepted one, so I would not classify this rule as a best practice. It might be useful as a coding convention, which is developer- or codebase-specific.

File: pmd-java/src/main/resources/rulesets/java/quickstart.xml [hunk 2]
```diff
@@ -51,6 +51,7 @@
     <rule ref="category/java/bestpractices.xml/UnusedPrivateField"/>
     <rule ref="category/java/bestpractices.xml/UnusedPrivateMethod"/>
     <rule ref="category/java/bestpractices.xml/UseCollectionIsEmpty"/>
+    <!-- <rule ref="category/java/bestpractices.xml/UseExplicitTypes"/> -->
```
Comment (adangel) [inline pmd-java/src/main/resources/rulesets/java/quickstart.xml:54]:
```suggestion
    <!-- <rule ref="category/java/codestyle.xml/UseExplicitTypes"/> -->
```

**Pull request merged**